File: java/src/test/java/com/ter/core/TypesTest.java
package com.ter.core;

import org.junit.Test;
import static org.junit.Assert.*;

public class TypesTest {
    @Test
    public void testStringType() {
        StringType type = Types.string();
        BaseType.ValidationResult result = type.validate("hello");
        assertTrue(result.valid);
        assertEquals("hello", result.value);
    }

    @Test
    public void testStringTypeCoercion() {
        StringType type = Types.string();
        BaseType.ValidationResult result = type.validate(123);
        assertTrue(result.valid);
        assertEquals("123", result.value);
    }

    @Test
    public void testStringTypeRequired() {
        StringType type = Types.string().markRequired();
        BaseType.ValidationResult result = type.validate(null);
        assertFalse(result.valid);
        assertTrue(result.error.contains("required"));
    }

    @Test
    public void testStringTypeMinLength() {
        StringType type = Types.string().minLength(5);
        assertFalse(type.validate("hi").valid);
        assertTrue(type.validate("hello").valid);
    }

    @Test
    public void testIntType() {
        IntType type = Types.integer();
        BaseType.ValidationResult result = type.validate(42);
        assertTrue(result.valid);
        assertEquals(42, result.value);
    }

    @Test
    public void testIntTypeCoercion() {
        IntType type = Types.integer();
        BaseType.ValidationResult result = type.validate("42");
        assertTrue(result.valid);
        assertEquals(42, result.value);
    }

    @Test
    public void testIntTypeMin() {
        IntType type = Types.integer().min(10);
        assertFalse(type.validate(5).valid);
        assertTrue(type.validate(15).valid);
    }

    @Test
    public void testNumberType() {
        NumberType type = Types.number();
        BaseType.ValidationResult result = type.validate(3.14);
        assertTrue(result.valid);
        assertEquals(3.14, result.value);
    }

    @Test
    public void testBooleanType() {
        BooleanType type = Types.bool();
        assertTrue(type.validate(true).valid);
        assertTrue(type.validate("true").valid);
        assertTrue(((Boolean) type.validate("false").value) == false);
    }

    @Test
    public void testEnumType() {
        EnumType type = Types.enumOf("dev", "prod", "staging");
        assertTrue(type.validate("dev").valid);
        assertFalse(type.validate("invalid").valid);
    }

    @Test
    public void testURLType() {
        URLType type = Types.url();
        assertTrue(type.validate("https://example.com").valid);
        assertFalse(type.validate("not a url").valid);
    }

    @Test
    public void testSecretType() {
        SecretType type = Types.secret();
        assertTrue(type.validate("secret_value").valid);
        assertTrue(type.isSecret());
    }
}

File: rust/tests/types_test.rs
use ter::core::{Types, ValidationResult};

#[test]
fn test_string_type() {
    let type_def = Types::string();
    match type_def.validate(Some("hello".to_string())) {
        ValidationResult::Valid(val) => {
            assert_eq!(val.as_str().unwrap(), "hello");
        }
        ValidationResult::Invalid(_) => panic!("Should be valid"),
    }
}

#[test]
fn test_string_required() {
    let type_def = Types::string().required();
    match type_def.validate(None) {
        ValidationResult::Invalid(e) => {
            assert!(e.contains("required"));
        }
        ValidationResult::Valid(_) => panic!("Should be invalid"),
    }
}

#[test]
fn test_int_type() {
    let type_def = Types::integer();
    match type_def.validate(Some("42".to_string())) {
        ValidationResult::Valid(val) => {
            assert_eq!(val.as_i64().unwrap(), 42);
        }
        ValidationResult::Invalid(_) => panic!("Should be valid"),
    }
}

#[test]
fn test_boolean_type() {
    let type_def = Types::boolean();
    match type_def.validate(Some("true".to_string())) {
        ValidationResult::Valid(val) => {
            assert_eq!(val.as_bool().unwrap(), true);
        }
        ValidationResult::Invalid(_) => panic!("Should be valid"),
    }
}

#[test]
fn test_enum_type() {
    let type_def = Types::enumof(vec!["dev".to_string(), "prod".to_string()]);
    match type_def.validate(Some("dev".to_string())) {
        ValidationResult::Valid(_) => {}
        ValidationResult::Invalid(_) => panic!("Should be valid"),
    }

    match type_def.validate(Some("invalid".to_string())) {
        ValidationResult::Invalid(_) => {}
        ValidationResult::Valid(_) => panic!("Should be invalid"),
    }
}

#[test]
fn test_secret_is_secret() {
    let type_def = Types::secret();
    assert!(type_def.is_secret());
}
